test/e2e/statefulset.go

```diff
@@ -247,44 +247,417 @@ var _ = framework.KubeDescribe("StatefulSet", func() {
+				Expect(pods.Items[i].Labels[apps.StatefulSetRevisionLabel]).To(Equal(currentRevision),
+					fmt.Sprintf("Pod %s/%s has revision %s not equal to current revision %s",
+						pods.Items[i].Namespace,
+						pods.Items[i].Name,
+						pods.Items[i].Labels[apps.StatefulSetRevisionLabel],
+						currentRevision))
+			}
+
+			By("By performing a canary update")
+			ss.Spec.UpdateStrategy = apps.StatefulSetUpdateStrategy{
+				Type: apps.RollingUpdateStatefulSetStrategyType,
+				RollingUpdate: func() *apps.RollingUpdateStatefulSetStrategy {
+					return &apps.RollingUpdateStatefulSetStrategy{
+						Partition: func() *int32 {
+							i := int32(2)
+							return &i
+						}()}
+				}(),
+			}
+			ss, err = framework.UpdateStatefulSetWithRetries(c, ns, ss.Name, func(update *apps.StatefulSet) {
+				update.Spec.UpdateStrategy = apps.StatefulSetUpdateStrategy{
+					Type: apps.RollingUpdateStatefulSetStrategyType,
+					RollingUpdate: func() *apps.RollingUpdateStatefulSetStrategy {
+						return &apps.RollingUpdateStatefulSetStrategy{
+							Partition: func() *int32 {
+								i := int32(2)
+								return &i
+							}()}
+					}(),
 				}
-				for i := range pods.Items {
-					if pods.Items[i].Spec.Containers[0].Image != newImage {
-						framework.Logf("Waiting for pod %s to have image %s current image %s",
+			})
+			Expect(err).NotTo(HaveOccurred())
+			ss, pods = sst.WaitForPartitionedRollingUpdate(ss)
+			for i := range pods.Items {
+				if i < int(*ss.Spec.UpdateStrategy.RollingUpdate.Partition) {
+					Expect(pods.Items[i].Spec.Containers[0].Image).To(Equal(oldImage),
+						fmt.Sprintf("Pod %s/%s has image %s not equal to current image %s",
+							pods.Items[i].Namespace,
 							pods.Items[i].Name,
-							newImage,
-							pods.Items[i].Spec.Containers[0].Image)
-						return false, nil
+							pods.Items[i].Spec.Containers[0].Image,
+							oldImage))
+					Expect(pods.Items[i].Labels[apps.StatefulSetRevisionLabel]).To(Equal(currentRevision),
+						fmt.Sprintf("Pod %s/%s has revision %s not equal to current revision %s",
+							pods.Items[i].Namespace,
+							pods.Items[i].Name,
+							pods.Items[i].Labels[apps.StatefulSetRevisionLabel],
+							currentRevision))
+				} else {
+					Expect(pods.Items[i].Spec.Containers[0].Image).To(Equal(newImage),
+						fmt.Sprintf("Pod %s/%s has image %s not equal to new image  %s",
+							pods.Items[i].Namespace,
+							pods.Items[i].Name,
+							pods.Items[i].Spec.Containers[0].Image,
+							newImage))
+					Expect(pods.Items[i].Labels[apps.StatefulSetRevisionLabel]).To(Equal(updateRevision),
+						fmt.Sprintf("Pod %s/%s has revision %s not equal to new revision %s",
+							pods.Items[i].Namespace,
+							pods.Items[i].Name,
+							pods.Items[i].Labels[apps.StatefulSetRevisionLabel],
+							updateRevision))
+				}
+			}
+
+			By("Restoring Pods to the correct revision when they are deleted")
+			sst.DeleteStatefulPodAtIndex(0, ss)
+			sst.DeleteStatefulPodAtIndex(2, ss)
```

Comment:
are these both in one side of partition or opposite sides ? worth putting a comment.